File: python/t1d_evidence/evidence.py
"""Load and validate the same evidence registry consumed by the web app."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def registry_path() -> Path:
    """Return the repository's canonical evidence-registry path."""
    return Path(__file__).resolve().parents[2] / "src" / "data" / "evidence-registry.json"


def load_registry(path: Path | None = None) -> list[dict[str, Any]]:
    """Load evidence records without changing their published values."""
    with (path or registry_path()).open(encoding="utf-8") as handle:
        data = json.load(handle)
    if not isinstance(data, list):
        raise ValueError("Evidence registry must contain a JSON array")
    return data


def evidence_by_id(evidence_id: str) -> dict[str, Any]:
    for record in load_registry():
        if record["id"] == evidence_id:
            return record
    raise KeyError(f"Unknown evidence id: {evidence_id}")


def validate_registry(records: list[dict[str, Any]] | None = None) -> list[str]:
    """Return every traceability error; an empty list means the registry is valid."""
    errors: list[str] = []
    seen: set[str] = set()
    required = {
        "id", "stage", "label", "estimateType", "point", "lower95", "upper95",
        "contrast", "horizon", "population", "sampleSize", "applicabilityWarning", "citation",
    }
    for index, item in enumerate(records or load_registry()):
        missing = sorted(required - item.keys())
        if missing:
            errors.append(f"record {index}: missing {', '.join(missing)}")
            continue
        evidence_id = str(item["id"])
        if evidence_id in seen:
            errors.append(f"duplicate evidence id: {evidence_id}")
        seen.add(evidence_id)
        if not item["lower95"] <= item["point"] <= item["upper95"]:
            errors.append(f"{evidence_id}: point lies outside its 95% CI")
        citation = item["citation"]
        if not citation.get("doi") or not str(citation.get("url", "")).startswith("https://"):
            errors.append(f"{evidence_id}: citation is incomplete")
        if not item["population"] or not item["applicabilityWarning"]:
            errors.append(f"{evidence_id}: population or applicability warning is missing")
    return errors

File: src/model/evidence.ts
import registryJson from '@/data/evidence-registry.json';
import type { EvidenceEstimate } from '@/types';

export const EVIDENCE_REGISTRY = registryJson as EvidenceEstimate[];

export function evidenceById(id: string): EvidenceEstimate {
  const evidence = EVIDENCE_REGISTRY.find((item) => item.id === id);
  if (!evidence) throw new Error(`Unknown evidence id: ${id}`);
  return evidence;
}

export function validateEvidenceRegistry(): string[] {
  const errors: string[] = [];
  const ids = new Set<string>();
  for (const item of EVIDENCE_REGISTRY) {
    if (ids.has(item.id)) errors.push(`Duplicate evidence id: ${item.id}`);
    ids.add(item.id);
    if (!(item.lower95 <= item.point && item.point <= item.upper95)) {
      errors.push(`${item.id}: point estimate is outside its 95% confidence range`);
    }
    if (!item.citation.doi || !item.citation.url) errors.push(`${item.id}: citation is incomplete`);
    if (!item.population || !item.applicabilityWarning) {
      errors.push(`${item.id}: population or applicability warning is missing`);
    }
  }
  return errors;
}
